pilot/pkg/model/push_context.go

```diff
@@ -967,33 +968,56 @@ func (ps *PushContext) getSidecarScope(proxy *Proxy, workloadLabels labels.Insta
 
 	// We didn't have a Sidecar in the namespace. This means we should use the default - either an implicit
 	// default selecting everything, or pulling from the root namespace.
-	ps.sidecarIndex.defaultSidecarMu.Lock()
-	defer ps.sidecarIndex.defaultSidecarMu.Unlock()
-	if proxy.Type == Router {
-		sc, f := ps.sidecarIndex.gatewayDefaultSidecarsByNamespace[proxy.ConfigNamespace]
+	var computed *SidecarScope
+	switch proxy.Type {
+	case Router:
+		ps.sidecarIndex.derivedSidecarMu.RLock()
+		// Gateways always use default sidecar scope.
+		sc, f := ps.sidecarIndex.catchAllSidecarsByNamespace[proxy.ConfigNamespace]
+		ps.sidecarIndex.derivedSidecarMu.RUnlock()
 		if f {
-			// We have already computed the scope for this namespace, just fetch it
+			// We have already computed the scope for this namespace, just return it.
 			return sc
 		}
-		computed := DefaultSidecarScopeForNamespace(ps, proxy.ConfigNamespace)
-		ps.sidecarIndex.gatewayDefaultSidecarsByNamespace[proxy.ConfigNamespace] = computed
-		return computed
-	}
-	sc, f := ps.sidecarIndex.computedSidecarsByNamespace[proxy.ConfigNamespace]
-	if f {
-		// We have already computed the scope for this namespace, just fetch it
-		return sc
+		// We need to compute this namespace
+		computed = ps.computeSidecarScope(proxy)
+	case SidecarProxy:
+		ps.sidecarIndex.derivedSidecarMu.RLock()
+		var sc *SidecarScope
+		var exists bool
+		if ps.sidecarIndex.meshRootSidecarConfig != nil {
+			sc, exists = ps.sidecarIndex.meshRootSidecarsByNamespace[proxy.ConfigNamespace]
+		} else {
+			sc, exists = ps.sidecarIndex.catchAllSidecarsByNamespace[proxy.ConfigNamespace]
+		}
+		ps.sidecarIndex.derivedSidecarMu.RUnlock()
+
+		if exists {
+			// We have already computed the scope for this namespace, just return it.
+			return sc
+		}
+		// We need to compute this namespace
+		computed = ps.computeSidecarScope(proxy)
 	}
-	// We need to compute this namespace
+	return computed
+}
+
+func (ps *PushContext) computeSidecarScope(proxy *Proxy) *SidecarScope {
+	ps.sidecarIndex.derivedSidecarMu.Lock()
+	defer ps.sidecarIndex.derivedSidecarMu.Unlock()
 	var computed *SidecarScope
-	if ps.sidecarIndex.rootConfig != nil {
-		computed = ConvertToSidecarScope(ps, ps.sidecarIndex.rootConfig, proxy.ConfigNamespace)
-	} else {
+	switch proxy.Type {
```

Comment:
same here. so no need to abstract another function, since this makes this lock spread across functions